Added expiration time to cached data

Tommaso-n opened the pull request:
Default value of expiration time set to 3600 seconds (1 hour).
If set to 0 or negative or None, data will never expire, otherwise set to the provided value in seconds

andreroggeri commented:
Hey @Tommaso-n I see that you're still making some changes on this PR. When you're done, please remove the WIP from the PR name so I can review.

Tommaso-n commented:
> Hey @Tommaso-n I see that you're still making some changes on this PR. When you're done, please remove the WIP from the PR name so I can review.

Hi!
Yes, sorry...now it should be definitive.

I've added a password for redis to the cached configuration (set to None by default), i use Redis Labs and it's required.

I've added two protected properties to the cached client:
- _redis_ttl is the expiration time of cached data, set to 3600 seconds by default.
- _redis_prefix is a string to be added as a prefix to the keys of the cached data for easier retrivial.

I've implemented the setter and getter of _redis_ttl.
If the new value is positive, set the property to the new value. If it's 0, negative or None, set the property to None (cached data never expire).

i've modified the get method to use the prefix for the keys.

To manually remove all the cached data relative to this package from cache, i've implemented a clear_cache method.
It scan for keys beginning with the given prefix and delete them.

This is all :)

andreroggeri commented on ynab_sdk/utils/clients/cached_client.py:
```diff
@@ -11,23 +12,45 @@ class CachedClient(BaseClient):
+        for k in self.redis.scan_iter(''.join([self._redis_prefix, '*'])):
+            self.redis.delete(k)
+            keys_count += 1
+        self.logger.error(f'clear_cache: {keys_count} keys deleted from cache')
```
Use logger.info since this is not an error

Tommaso-n replied:
Done

andreroggeri commented on ynab_sdk/utils/clients/cached_client.py:
```diff
@@ -11,23 +12,45 @@ class CachedClient(BaseClient):
+    @cache_time_to_live.setter
+    def cache_time_to_live(self, ttl_value: Union[int, None]) -> None:
+        if ttl_value and ttl_value > 0:
+            self.logger.error(f'cache_time_to_live: TTL value set to {ttl_value}')
```
Use logger.debug since this is not an error

Tommaso-n replied:
Done

andreroggeri commented on ynab_sdk/utils/clients/cached_client.py:
```diff
@@ -11,23 +12,45 @@ class CachedClient(BaseClient):
+            self.logger.error(f'cache_time_to_live: TTL value set to {ttl_value}')
+            self._redis_ttl = ttl_value
+        else:
+            self.logger.error(f'cache_time_to_live: 0, negative or None TTL value: {ttl_value}, set to None')
```
Use logger.debug since this is not an error

Tommaso-n replied:
Done

andreroggeri commented on ynab_sdk/utils/clients/cached_client.py:
```diff
@@ -11,23 +12,45 @@ class CachedClient(BaseClient):
         super().__init__(config)
-        self.redis = Redis(host=config.redis_host, port=config.redis_port, db=config.redis_db)
+        self.redis = Redis(host=config.redis_host, port=config.redis_port, db=config.redis_db, password=config.redis_pass)
+        self._redis_ttl: Union[int, None] = 3600
```
Optional[int] its the correct typing here (Has the same meaning, but a matter of semantics)

Tommaso-n replied:
Done

andreroggeri commented on ynab_sdk/utils/configurations/cached.py:
```diff
@@ -3,8 +3,9 @@
 class CachedConfig(DefaultConfig):
 
-    def __init__(self, redis_host: str, redis_port: int, redis_db: int = 0, **kwargs):
+    def __init__(self, redis_host: str, redis_port: int, redis_db: int = 0, redis_pass: str = None, **kwargs) -> None:
```
`redis_pass` should be typed as Optional[str]

Tommaso-n replied:
Done

andreroggeri commented on ynab_sdk/utils/clients/cached_client.py:
```diff
@@ -11,23 +12,45 @@ class CachedClient(BaseClient):
         super().__init__(config)
-        self.redis = Redis(host=config.redis_host, port=config.redis_port, db=config.redis_db)
+        self.redis = Redis(host=config.redis_host, port=config.redis_port, db=config.redis_db, password=config.redis_pass)
+        self._redis_ttl: Union[int, None] = 3600
```
Hmm, maybe the ttl should be a property of the CachedConfig.
Its a little odd that after all configuration I still have to call another config method

Tommaso-n replied:
Moved

andreroggeri commented on ynab_sdk/utils/clients/cached_client.py:
```diff
@@ -11,23 +12,45 @@ class CachedClient(BaseClient):
+        return self._redis_ttl
+
+    @cache_time_to_live.setter
+    def cache_time_to_live(self, ttl_value: Union[int, None]) -> None:
```
`ttl_value`  should be Optional[int]

Tommaso-n replied:
Done

andreroggeri commented on ynab_sdk/utils/configurations/cached.py:
```diff
@@ -1,10 +1,27 @@
+        self._redis_ttl: Optional[int] = 3600
+
+    @property
+    def redis_ttl(self) -> Union[int, None]:
```
I missed this one: the return type is Optional[int]

Tommaso-n replied:
I missed it too...corrected
